efi/pe.go

```diff
@@ -0,0 +1,261 @@
+}
+
+func (h *peImageHandleImpl) SbatComponents() ([]sbatComponent, error) {
+	sr := h.OpenSection(".sbat")
+	if sr == nil {
+		return nil, errors.New("no .sbat section")
+	}
+
+	r := csv.NewReader(newCstringReader(sr))
+	r.FieldsPerRecord = 6
+
+	var components []sbatComponent
+	for {
+		record, err := r.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, xerrors.Errorf("invalid record: %w", err)
+		}
+
+		gen, err := strconv.Atoi(record[1])
+		if err != nil {
+			return nil, xerrors.Errorf("invalid component generation: %w", err)
+		}
+		component := sbatComponent{
+			Name:              record[0],
+			Generation:        gen,
+			VendorName:        record[2],
+			VendorPackageName: record[3],
+			VendorVersion:     record[4],
+			VendorUrl:         record[5]}
+
+		if component.Name == "sbat" {
+			if component.Generation != 1 {
+				return nil, xerrors.Errorf("invalid .sbat section version")
+			}
+			continue
+		}
+
+		components = append(components, component)
+	}
+
+	return components, nil
+}
+
+const (
+	certTableIndex = 4 // Index of the Certificate Table entry in the Data Directory of a PE image optional header
+)
+
+func (h *peImageHandleImpl) ImageDigest(alg crypto.Hash) ([]byte, error) {
+	return efi.ComputePeImageDigest(alg, h.r, h.r.Size())
+}
+
+func (h *peImageHandleImpl) SecureBootSignatures() ([]*efi.WinCertificateAuthenticode, error) {
+	// Obtain security directory entry from optional header
+	var dd []pe.DataDirectory
+	switch oh := h.pefile.OptionalHeader.(type) {
+	case *pe.OptionalHeader32:
+		dd = oh.DataDirectory[0:oh.NumberOfRvaAndSizes]
+	case *pe.OptionalHeader64:
+		dd = oh.DataDirectory[0:oh.NumberOfRvaAndSizes]
+	default:
+		return nil, errors.New("cannot obtain security directory entry: no optional header")
+	}
+
+	if len(dd) <= certTableIndex {
+		// This image doesn't include a certificate table entry, so has no signatures.
+		return nil, nil
+	}
+
+	// Create a reader for the security directory entry, which points to one or more WIN_CERTIFICATE structs
+	certReader := io.NewSectionReader(
+		h.r,
+		int64(dd[certTableIndex].VirtualAddress),
+		int64(dd[certTableIndex].Size))
+
+	// Binaries can have multiple signers - this is achieved using multiple single-signed Authenticode
+	// signatures - see section 32.5.3.3 ("Secure Boot and Driver Signing - UEFI Image Validation -
+	// Signature Database Update - Authorization Process") of the UEFI Specification, version 2.8.
+	var sigs []*efi.WinCertificateAuthenticode
+
+	noMoreCerts := false
+	for i := 0; ; i++ {
+		// Signatures in this section are 8-byte aligned - see the PE spec:
+		// https://docs.microsoft.com/en-us/windows/win32/debug/pe-format#the-attribute-certificate-table-image-only
+		off, _ := certReader.Seek(0, io.SeekCurrent)
+		alignSize := (8 - (off & 7)) % 8
+		certReader.Seek(alignSize, io.SeekCurrent)
+
+		c, err := efi.ReadWinCertificate(certReader)
+		switch {
+		case xerrors.Is(err, io.EOF):
+			noMoreCerts = true
```

Comment:
this could use a "break Label" statement instead of a var

Reply:
I've changed this to use a label.